### ISSUE TITLE
Parsing issues with Chemical Environments

### ISSUE THREAD

**Issue by bannanc** [open-forcefield-group/openforcefield#16]
There are a few problems with chemical environments when it comes to substituted SMIRKS strings, such as `"$ewg1"`. I wrote up long commentary in the smarty issue tracker before remembering we're trying to move everything to this repository, please see [smarty issue#248](https://github.com/open-forcefield-group/smarty/issues/248)

**Comment by bannanc** [comment]
Here's the brief version of what is wrong:

The other parsing issue has to do with environment assuming all atoms are specified with square brackets. However, SMIRKS/SMARTS build on smiles strings so they can also use elemental symbols. For example, this is a reasonable smirks string for an H1:

[#1:1]-C-$ewg1

Right now (before pull request) the environment would look for atoms inside square brackets [stuff] and would only find the hydrogen and would loose all the other information. 

I am adding some sort of weird RegEx patterns that fix this problem so that it can correctly parse the carbon and $ewg1. It still has a few bugs, but I should be able to get them ironed out in the morning. 

------------------------

The other issue is that we have allowed users to specify blank OR or AND decorators as ''. This makes sampling possible decorator options easier, but those blanks need to be removed inside environments. The non-parseable smirks from SMIRKY had the pattern `[$ewg1&]` the `&` was added in order to add ORdecorators to the replacement string $ewg1, but the only decorator was a blank ''. I'm adding more thorough checks that remove blanks and repeating decorators from all user input places.
